_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+
+This standard approach works poorly in Rust, though it can be done (see in particular the [Dioxus] system for an example of one of the most literal transliterations of React patterns, including observer-based state updating, into Rust). The problem is that it requires *shared mutable* access to that state, which is clunky at best in Rust (it requires interior mutability). In addition, because Rust doesn't have built-in syntax for getters and setters, invoking the notification mechanism also requires some kind of explicit call (though perhaps macros or other techniques can be used to hide it or make it less prominent).
+
+<details>
+<summary>Comparison with Elm</summary>
+
+The observer pattern is not the *only* way event propagation works in declarative UI. Another very important and influential pattern is [The Elm Architecture], which, being based on a pure functional language, also does not require shared mutable state. Thus, it is also used successfully as the basis of several Rust UI toolkits, notably Iced.
+
+In Elm, app state is centralized (this is also a fairly popular pattern in React, using state management packages such as Redux), and events are given to the app through an `update` call. Dispatching is a three-stage process. First, the user defines a *message* type enumerating the various actions that are (globally) possible to trigger through the UI. Second, the UI element *maps* the event type into this user-defined type, identifying which action is desired. Third, the `update` method dispatches the event, delegating if needed to a child handler. Some people like the explicitness of this approach, but it is unquestionably more verbose than a single callback that manipulates state directly as in React or SwiftUI.
+</details>
+
+So what does Xilem do instead? The view tree is also parameterized on the *app state,* which can be any type. This idea is an evolution of Druid's existing architecture, which also offers mutable access to app state to UI callbacks, but removes some of the limitations. In particular, Druid requires app state to be clonable and diffable, a stumbling block for many new users.
+
+When an event is generated, it is annotated with the path of the UI element that originated it. In the case of the button, `[1, 3]`. We retain the view tree for a fairly short time, but long enough to do event propagation. The event is dispatched from the *root,* and each successive node strips off one id from the id path, choosing a child, and traversing to that child. In this case, that's simple, the enclosing stack just chooses the button. This event propagation also passes down mutable access to the app state.
+
+Given that setup, running the closure is fairly straightforward. The event is dispatched to the stack, which uses the id to recognize that it needs to be routed to its second child, and that button is the target. It runs the callback on the mutable app state, which in this case just increments `count`.
```

Comment:
Even after reading this I'm still confused how the vstack knows to route the event to the button and how this also provides mutable access. Also, what event is this? Is this something like a mouse button event?

I think this part could do with some more explanation.